core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java

```diff
@@ -0,0 +1,1668 @@
+   }
+
+   @Override
+   public NotifyingFuture<V> putIfAbsentAsync(K key, V value, long lifespan, TimeUnit unit) {
+      return new NoOpFuture<>(putIfAbsent(key, value, lifespan, unit));
+   }
+
+   @Override
+   public NotifyingFuture<V> putIfAbsentAsync(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdle, TimeUnit maxIdleUnit) {
+      return new NoOpFuture<>(putIfAbsent(key, value, lifespan, lifespanUnit, maxIdle, maxIdleUnit));
+   }
+
+   @Override
+   public NotifyingFuture<V> removeAsync(Object key) {
+      return new NoOpFuture<>(remove(key));
+   }
+
+   @Override
+   public NotifyingFuture<Boolean> removeAsync(Object key, Object value) {
+      return new NoOpFuture<>(remove(key, value));
+   }
+
+   @Override
+   public NotifyingFuture<V> replaceAsync(K key, V value) {
+      return new NoOpFuture<>(replace(key, value));
+   }
+
+   @Override
+   public NotifyingFuture<V> replaceAsync(K key, V value, long lifespan, TimeUnit unit) {
+      return new NoOpFuture<>(replace(key, value, lifespan, unit));
+   }
+
+   @Override
+   public NotifyingFuture<V> replaceAsync(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdle, TimeUnit maxIdleUnit) {
+      return new NoOpFuture<>(replace(key, value, lifespan, lifespanUnit, maxIdle, maxIdleUnit));
+   }
+
+   @Override
+   public NotifyingFuture<Boolean> replaceAsync(K key, V oldValue, V newValue) {
+      return new NoOpFuture<>(replace(key, oldValue, newValue));
+   }
+
+   @Override
+   public NotifyingFuture<Boolean> replaceAsync(K key, V oldValue, V newValue, long lifespan, TimeUnit unit) {
+      return new NoOpFuture<>(replace(key, oldValue, newValue, lifespan, unit));
+   }
+
+   @Override
+   public NotifyingFuture<Boolean> replaceAsync(K key, V oldValue, V newValue, long lifespan, TimeUnit lifespanUnit, long maxIdle, TimeUnit maxIdleUnit) {
+      return new NoOpFuture<>(replace(key, oldValue, newValue, lifespan, lifespanUnit, maxIdle, maxIdleUnit));
+   }
+
+   @Override
+   public NotifyingFuture<V> getAsync(K key) {
+      return new NoOpFuture<>(get(key));
+   }
+
+   @Override
+   public boolean startBatch() {
+      // invocation batching implies CacheImpl
+      throw log.invocationBatchingNotEnabled();
+   }
+
+   @Override
+   public void endBatch(boolean successful) {
+      // invocation batching implies CacheImpl
+      throw log.invocationBatchingNotEnabled();
+   }
+
+   @Override
+   public V putIfAbsent(K key, V value) {
+      return putIfAbsentInternal(key, value, defaultMetadata);
+   }
+
+   @Override
+   public boolean remove(Object key, Object value) {
+      Objects.requireNonNull(key, NULL_KEYS_NOT_SUPPORTED);
+      Objects.requireNonNull(value, NULL_VALUES_NOT_SUPPORTED);
+      ByRef<InternalCacheEntry<K, V>> oldEntryRef = new ByRef<>(null);
+      boolean hasListeners = this.hasListeners;
+      // TODO: we will probably hit runtime error here if key is not of type K, but we can't even do a check here
```

Comment:
Exactly, `Cache` isn't narrowing that, so throwing an exception at runtime isn't ok :)

Actually, I don't expect any `ClassCastException` here, because `K` is erased both in the `Cache` and in the `DataContainer`. But you should definitely have a test for it (and remove the TODO).